File: Sources/custom_adapter/UnityLevelPlayAdapter/Objc/ISAlxCustomRewardedVideo.m
//
//  ISAlxCustomRewardedVideo.m
//  AlxAdsOCDemo
//
//  LevelPlay RewardedVideo 广告适配器 / LevelPlay RewardedVideo ad adapter
//

#import "ISAlxCustomRewardedVideo.h"
#import <AlxAds/AlxAds-Swift.h>

static NSString *const TAG = @"ISAlxCustomRewardedVideo";

@interface ISAlxCustomRewardedVideo () <AlxRewardVideoAdDelegate>
@property (nonatomic, strong, nullable) AlxRewardVideoAd *rewardedAd;
@property (nonatomic, weak, nullable) id<ISRewardedVideoAdDelegate> adDelegate;
@end

@implementation ISAlxCustomRewardedVideo

#pragma mark - ISBaseRewardedVideo

/**
 * LevelPlay 请求加载激励视频广告。
 * LevelPlay requests to load rewarded video ad.
 */
- (void)loadAdWithAdData:(ISAdData *)adData delegate:(id<ISRewardedVideoAdDelegate>)delegate {
    NSLog(@"%@: loadAd", TAG);
    self.adDelegate = delegate;

    [ISAlxCustomAdapter initSdkWithAdData:adData];

    NSString *unitId = adData.configuration[@"unitid"];
    if (!unitId || unitId.length == 0) {
        NSString *msg = @"unitid is empty";
        NSLog(@"%@: error: %@", TAG, msg);
        [delegate adDidFailToLoadWithErrorType:ISAdapterErrorTypeInternal
                                     errorCode:ISAdapterErrorMissingParams
                                  errorMessage:msg];
        return;
    }

    NSLog(@"%@: loadAd unitid=%@", TAG, unitId);

    self.rewardedAd = [[AlxRewardVideoAd alloc] init];
    self.rewardedAd.delegate = self;
    [self.rewardedAd loadAdWithAdUnitId:unitId];
}

/**
 * LevelPlay 检查广告是否就绪。
 * LevelPlay checks if the ad is ready.
 */
- (BOOL)isAdAvailableWithAdData:(ISAdData *)adData {
    BOOL ready = self.rewardedAd != nil && [self.rewardedAd isReady];
    NSLog(@"%@: isAdAvailable = %@", TAG, ready ? @"YES" : @"NO");
    return ready;
}

/**
 * LevelPlay 展示激励视频广告。
 * LevelPlay shows rewarded video ad.
 */
- (void)showAdWithViewController:(UIViewController *)viewController
                          adData:(ISAdData *)adData
                        delegate:(id<ISRewardedVideoAdDelegate>)delegate {
    NSLog(@"%@: showAd", TAG);
    self.adDelegate = delegate;

    if (![self isAdAvailableWithAdData:adData]) {
        NSString *msg = @"ad is not ready";
        NSLog(@"%@: error: %@", TAG, msg);
        [delegate adDidFailToShowWithErrorCode:ISAdapterErrorInternal errorMessage:msg];
        return;
    }

    dispatch_async(dispatch_get_main_queue(), ^{
        [self.rewardedAd showAdWithPresent:viewController];
    });
}

#pragma mark - AlxRewardVideoAdDelegate

- (void)rewardVideoAdLoad:(AlxRewardVideoAd *)ad {
    NSLog(@"%@: rewardVideoAdLoad", TAG);
    [self.adDelegate adDidLoad];
}

- (void)rewardVideoAdFailToLoad:(AlxRewardVideoAd *)ad didFailWithError:(NSError *)error {
    NSLog(@"%@: rewardVideoAdFailToLoad: %@", TAG, error.localizedDescription);
    [self.adDelegate adDidFailToLoadWithErrorType:ISAdapterErrorTypeInternal
                                        errorCode:error.code
                                     errorMessage:error.localizedDescription];
}

- (void)rewardVideoAdImpression:(AlxRewardVideoAd *)ad {
    NSLog(@"%@: rewardVideoAdImpression", TAG);
    [self.adDelegate adDidOpen];
    [self.adDelegate adDidBecomeVisible];
}

- (void)rewardVideoAdClick:(AlxRewardVideoAd *)ad {
    NSLog(@"%@: rewardVideoAdClick", TAG);
    [self.adDelegate adDidClick];
}

- (void)rewardVideoAdClose:(AlxRewardVideoAd *)ad {
    NSLog(@"%@: rewardVideoAdClose", TAG);
    [self.adDelegate adDidClose];
    self.rewardedAd = nil;
}

- (void)rewardVideoAdPlayStart:(AlxRewardVideoAd *)ad {
    NSLog(@"%@: rewardVideoAdPlayStart", TAG);
    [self.adDelegate adDidStart];
}

- (void)rewardVideoAdPlayEnd:(AlxRewardVideoAd *)ad {
    NSLog(@"%@: rewardVideoAdPlayEnd", TAG);
    [self.adDelegate adDidEnd];
}

- (void)rewardVideoAdReward:(AlxRewardVideoAd *)ad {
    NSLog(@"%@: rewardVideoAdReward", TAG);
    [self.adDelegate adRewarded];
}

- (void)rewardVideoAdPlayFail:(AlxRewardVideoAd *)ad didFailWithError:(NSError *)error {
    NSLog(@"%@: rewardVideoAdPlayFail: %@", TAG, error.localizedDescription);
    [self.adDelegate adDidFailToShowWithErrorCode:ISAdapterErrorInternal
                                     errorMessage:error.localizedDescription];
}

@end

File: custom_adapter/UnityLevelPlayAdapter/Swift/ISAlxCustomRewardedVideo.swift
//
//  ISAlxCustomRewardedVideo.swift
//  AlxAdsDemo
//
//  LevelPlay RewardedVideo 广告适配器 / LevelPlay RewardedVideo Ad Adapter
//  文档参考 / Documentation: https://docs.unity.com/zh-cn/grow/levelplay/sdk/ios/build-custom-adapter
//

import Foundation
import UIKit
import IronSource
import AlxAds

@objc(ISAlxCustomRewardedVideo)
public class ISAlxCustomRewardedVideo: ISBaseRewardedVideo {

    private static let TAG = "ISAlxCustomRewardedVideo"

    private var rewardedAd: AlxRewardVideoAd?
    private weak var adDelegate: ISRewardedVideoAdDelegate?

    // MARK: - ISBaseRewardedVideo

    public override func loadAd(with adData: ISAdData, delegate: ISRewardedVideoAdDelegate) {
        NSLog("%@: loadAd", Self.TAG)
        self.adDelegate = delegate
        ISAlxCustomAdapter.initSdk(with: adData)

        guard let unitId = adData.configuration["unitid"] as? String, !unitId.isEmpty else {
            let msg = "unitid is empty"
            NSLog("%@: error: %@", Self.TAG, msg)
            (delegate as? ISAlxAdapterFailureReporter)?.adDidFailToLoad(
                withErrorType: 2,
                errorCode: ISAdapterErrors.missingParams.rawValue,
                errorMessage: msg)
            return
        }

        NSLog("%@: loadAd unitid=%@", Self.TAG, unitId)
        let ad = AlxRewardVideoAd()
        ad.delegate = self
        self.rewardedAd = ad
        ad.loadAd(adUnitId: unitId)
    }

    public override func isAdAvailable(with adData: ISAdData) -> Bool {
        let ready = rewardedAd?.isReady() ?? false
        NSLog("%@: isAdAvailable = %@", Self.TAG, ready ? "YES" : "NO")
        return ready
    }

    public override func showAd(with viewController: UIViewController,
                                adData: ISAdData,
                                delegate: ISRewardedVideoAdDelegate) {
        NSLog("%@: showAd", Self.TAG)
        self.adDelegate = delegate

        guard isAdAvailable(with: adData) else {
            let msg = "ad is not ready"
            NSLog("%@: error: %@", Self.TAG, msg)
            (delegate as? ISAlxAdapterFailureReporter)?.adDidFailToShow(
                withErrorCode: 1000,
                errorMessage: msg)
            return
        }

        DispatchQueue.main.async { [weak self] in
            self?.rewardedAd?.showAd(present: viewController)
        }
    }
}

// MARK: - AlxRewardVideoAdDelegate

extension ISAlxCustomRewardedVideo: AlxRewardVideoAdDelegate {

    public func rewardVideoAdLoad(_ ad: AlxRewardVideoAd) {
        NSLog("%@: rewardVideoAdLoad", Self.TAG)
        adDelegate?.adDidLoad()
    }

    public func rewardVideoAdFailToLoad(_ ad: AlxRewardVideoAd, didFailWithError error: Error) {
        NSLog("%@: rewardVideoAdFailToLoad: %@", Self.TAG, error.localizedDescription)
        (adDelegate as? ISAlxAdapterFailureReporter)?.adDidFailToLoad(
            withErrorType: 2,
            errorCode: (error as NSError).code,
            errorMessage: error.localizedDescription)
    }

    public func rewardVideoAdImpression(_ ad: AlxRewardVideoAd) {
        NSLog("%@: rewardVideoAdImpression", Self.TAG)
        adDelegate?.adDidOpen()
        adDelegate?.adDidBecomeVisible()
    }

    public func rewardVideoAdClick(_ ad: AlxRewardVideoAd) {
        NSLog("%@: rewardVideoAdClick", Self.TAG)
        adDelegate?.adDidClick()
    }

    public func rewardVideoAdClose(_ ad: AlxRewardVideoAd) {
        NSLog("%@: rewardVideoAdClose", Self.TAG)
        adDelegate?.adDidClose()
        rewardedAd = nil
    }

    public func rewardVideoAdPlayStart(_ ad: AlxRewardVideoAd) {
        NSLog("%@: rewardVideoAdPlayStart", Self.TAG)
        adDelegate?.adDidStart()
    }

    public func rewardVideoAdPlayEnd(_ ad: AlxRewardVideoAd) {
        NSLog("%@: rewardVideoAdPlayEnd", Self.TAG)
        adDelegate?.adDidEnd()
    }

    public func rewardVideoAdReward(_ ad: AlxRewardVideoAd) {
        NSLog("%@: rewardVideoAdReward", Self.TAG)
        adDelegate?.adRewarded()
    }

    public func rewardVideoAdPlayFail(_ ad: AlxRewardVideoAd, didFailWithError error: Error) {
        NSLog("%@: rewardVideoAdPlayFail: %@", Self.TAG, error.localizedDescription)
        (adDelegate as? ISAlxAdapterFailureReporter)?.adDidFailToShow(
            withErrorCode: 1000,
            errorMessage: error.localizedDescription)
    }
}
